### ISSUE TITLE
Initialize default group list via providers

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#13831]
Problem statement: Currently, only internal(ly implemented) groups can be TLS default groups as per this static, "manually prepared" list: https://github.com/openssl/openssl/blob/e211d949cd5737e53cd3399e6a88453930768b98/ssl/t1_lib.c#L196-L223 (Runtime) changes to this list require use of `-groups` (CLI) or `SSL_set1_groups_list` (API) calls. 

Changing this way can be tedious (in testing) or error-prone (if working with external provider groups). Also, in case of absence of (these) groups (e.g., if EC+DH are not present as discussed [here](https://github.com/openssl/openssl/issues/13767)) this list can turn empty even though providers may have been registered with alternative groups (e.g., [tls-provider](https://github.com/openssl/openssl/blob/master/test/tls-provider.c)). 

The proposal is to make the currently static data structure `supported_groups_default` dynamic and populate it with group IDs suitably tagged when loading all providers [here](https://github.com/openssl/openssl/blob/e211d949cd5737e53cd3399e6a88453930768b98/ssl/t1_lib.c#L239). This may be done by adding the flag `default_group` to this data structure: https://github.com/openssl/openssl/blob/e211d949cd5737e53cd3399e6a88453930768b98/ssl/ssl_local.h#L812-L822 and evaluating it when loading the groups from providers.

I'd volunteer adding this logic if this feature is deemed desirable. Effort seems not overly big.

**Comment by romen** [comment]
I am not sure about this feature: runtime selection of the groups is also possible via conf file.
For both testing and the installation of providers I am not convinced we want the provider to potentially alter the configuration file: that should be the central policy registry, at least in my current understanding.

Maybe @mattcaswell can chime in about it.

Don't take this as a firm "no", as I admit that libssl is not my main area of expertise.

**Comment by baentsch** [comment]
Good point(er): Thanks. Maybe all that's needed is a modified test.cnf file to pass a `-no-ec -no-dh` config testsuite. Let me check that out.

**Comment by romen** [comment]
Just to expand on this, in my current understanding, in the scenario of a user/sysadmin installing a OQS-provider for OpenSSL 3.0, the idea would be:

- download a pre-compiled binary or build the provider from source
- install the resulting loadable module in `MODULESDIR` (as reported by `openssl version -m` or as configured per-application using the `OPENSSL_MODULES` env var)
- edit the global `OPENSSLDIR/openssl.cnf` (`openssl version -d`) or the per-application configuration file set using the `OPENSSL_CONF` env var to:
  - load the OQS provider
  - add whatever KEM the user wants to support by default to the default ssl configuration for server and/or client
  
For reliable testing, we could do this by making sure the test app overrides the default groups adding the XORKEM/XORKEX to the default list or loading an ad-hoc config file.

**Comment by mattcaswell** [comment]
> add whatever KEM the user wants to support by default to the default ssl configuration for server and/or client

One issue I see with this is that it is not possible to just *add* new groups to the default list. You have to configure the whole list. That of course requires you to know what the default was before.

**Comment by t8m** [comment]
I've originally misunderstood the @baentsch proposal. Thinking about it more, it makes actually some sense. And of course the default groups from loaded providers would take an effect only in case the user does not explicitly configure the list.

**Comment by mattcaswell** [comment]
> For both testing and the installation of providers I am not convinced we want the provider to potentially alter the configuration as set by the configuration file

I don't think this is what was proposed. This is just about what is in the *default* list. As @t8m says above it would only take effect if the user doesn't explicitly configure something different via the config file, or via API.

**Comment by romen** [comment]
Ok, thank you both @mattcaswell and @t8m, you both have very good points, and my understanding was limited since the beginning!

So I can now see the point of appending to the default list (the one that is currently static) from the providers: this list is overridden by setting any preference for supported groups either from the conf file or from application calls.
During the tests this would allow a default list to be populated with at least the xor groups if no FF/EC groups are in the default list (for `no-ec no-dh` builds).

But, for regular use with a complete build, a user that wanted to use the OQS provider and prefer post-quantum crypto over X25519 or P-256, the config file would still need to be overwritten, as in that case the user also needs to reorder preference (especially for the client configuration) not just append. Is my understanding correct?

**Comment by mattcaswell** [comment]
> But, for regular use with a complete build, a user that wanted to use the OQS provider and prefer post-quantum crypto over X25519 or P-256, they would need to edit the supported groups in the config file, as in that case they also need to reorder preference (especially for the client configuration) not just append. Is my understanding correct?

Actually this point does throw up an issue with this proposal. The way I envisage this working is that we extend the TLS-GROUPS capability with additional info about whether a group is in the default list or not. That, presumably, would apply to the built in groups too. But where does the information about the default *ordering* come from?

**Comment by baentsch** [comment]
> One issue I see with this is that it is not possible to just add new groups to the default list. You have to configure the whole list. That of course requires you to know what the default was before.

Just _adding_ (new, possibly "default-declared") groups was my original thought. I then wondered whether it wouldn't be beneficial/in line with the provider concept to give more control over this to the provider (not just OQS, but also FIPS et.al.). 

> Actually this point does throw up an issue with this proposal. The way I envisage this working is that we extend the TLS-GROUPS capability with additional info about whether a group is in the default list or not. That, presumably, would apply to the built in groups too. But where does the information about the default ordering come from?

That is a very valid concern. What about replacing the original(ly proposed) `default_group` bit with a `default_group_priority` indicator (say `uint8_t`) that the provider(s) can set? For example, 0 could mean, don't include in default list, 1 include at the end, ...255 place at the top of the list (or the like). Alternatively, "internal provider" groups' defaults go first, followed by external providers' defaults. At the very end, the order can always be set via `-groups`/`SSL_set1_groups_list()`.

**Comment by richsalz** [comment]
>  1 include at the end, ...255 place at the top of the list (or the like).

That seems like a bad idea, because what actually happens will depend on the order in which providers are loaded.  Note that the order isn't defined in the documentation.

**Comment by baentsch** [comment]
> because what actually happens will depend on the order in which providers are loaded

That would be the case only if different providers use identical priority numbers; if those priority numbers are strictly separate across providers the order would be deterministic, regardless of when each provider is loaded: Its groups will always be sorted into the default group list depending on their priority. So, if, say, X25519 from the FIPS provider would state priority number "250" and no other provider contributes a group with a higher priority, X25519 would get at the top of the list.

**Comment by richsalz** [comment]
> That would be the case only if different providers use identical priority numbers; 

Yes.  It seems more than a little embarassing for a cryptographic library to work as if eight bits will never have conflicts. :)

>the FIPS provider would state priority 250

Which FIPS provider?  What about third-parties?  Platforms?

**Comment by mattcaswell** [comment]
> Yes. It seems more than a little embarassing for a cryptographic library to work as if eight bits will never have conflicts. :)

I'd say there's no reason to restrict ourselves to a small space. I'd have though at least we should use at least 16bits - or possibly even 32 or 64.

> Which FIPS provider? What about third-parties? Platforms?

If we go with this approach then I'd suggest we "reserve" a set of priorities in the middle of the range for OpenSSL's own providers. Outside of that range (either "below" or "above") then external providers can choose whatever priorities they see fit. The only limitation would then be if a provider wanted to insert a group "in the middle" of our priority list i.e. "after group x but before group y".

**Comment by baentsch** [comment]
> Yes. It seems more than a little embarassing for a cryptographic library to work as if eight bits will never have conflicts. :)

Well, even in 8 bits one could parcel out (and enforce/protect during load) specific ranges for special (say internal) providers; but then again one may splurge on bits and use a real `int` :-) Fun aside: I did not find a clear-cut spec location defining a specific default group ordering in RFC 8446, 8422 and 7919; can you point to what drives/defines this preference ordering?

**Comment by richsalz** [comment]
> Fun aside: I did not find a clear-cut spec location defining a specific default group ordering in RFC 8446, 8422 and 7919; can you point to what drives/defines this preference ordering?

There is no official ordering, it is purely up to the application to decide.

In that view, perhaps the best thing to do is to "providers append their groups" and applications that care can `SSL_set1_groups_list`

**Comment by mattcaswell** [comment]
> can you point to what drives/defines this preference ordering?

We discuss it amongst ourselves and decide the ordering we think based on our knowledge.

**Comment by mattcaswell** [comment]
> In that view, perhaps the best thing to do is to "providers append their groups"

That might work as long as we can identify which groups have come from a third party provider, and which have come from our own providers.

**Comment by t8m** [comment]
Yeah, I am afraid with the priorities we would over-engineer this without making it fully working for all use-cases anyway. So perhaps we should just initialize the list with what we currently have and let other providers append things in the order they are loaded. (Of course if a provider adds something that is already there, it is not added.)

**Comment by t8m** [comment]
And for anything else the user can configure stuff explicitly.

**Comment by mattcaswell** [comment]
WFM.

**Comment by mattcaswell** [comment]
WFM = Works For Me. I now discover that according to urban dictionary that WFM can mean "Works For Me", but can also mean a whole load of other things that I most definitely did *not* mean :-)

https://www.urbandictionary.com/define.php?term=wfm

**Comment by baentsch** [comment]
> So perhaps we should just initialize the list with what we currently have and let other providers append things in the order they are loaded

What if someone wants to configure OpenSSL without "internal" (default, FIPS) providers? In that case pre-seeding the "list we currently have" would be incorrect: X25519 etc are listed but wouldn't be available, really.

**Comment by t8m** [comment]
I should have been more specific - with the "pre-seeding" I meant for the purpose of ordering the initial default groups. We should just mark the actually available when the providers are queried. So basically the mechanism for providing the default groups would be the same for external and internal providers just the groups from the internal providers would have this special advantage that they are sorted first and in a fixed order.

**Comment by baentsch** [comment]
So the pseudo code to populate `supported_groups_default` should then look like this:
1) For each provider loaded,
2) for each group in provider,
3) check whether it has `default_group` bit set (new code in [this place](https://github.com/openssl/openssl/blob/0d11846e4b2850773d1ee0df206608549a7d45d0/ssl/t1_lib.c#L373)) and if so,
3a) If group is in "internal pre-seed list" (i.e., the current static list), insert it into (the proposed new dynamic) `supported_groups_default` list at its (relative) position between other already registered "pre-seed list" groups
3b) If group is not in "internal pre-seed list", simply append it to `supported_groups_default`

This would (still) necessitate adding the `default_group` bit to `TLS_GROUP_INFO` also in the internal providers. Do I understand you right? Or do you suggest inserting any group on the "preseed list" regardless of whether or not it has a `default_group` bit set (as it is a "privileged prioritized group" to begin with)?

This seems to be getting way too complicated and hard to document/explain now, compared to the current "if the [currently built-in choice and priority of default groups](https://github.com/openssl/openssl/blob/0d11846e4b2850773d1ee0df206608549a7d45d0/ssl/t1_lib.c#L196) doesn't work for you, use `SSL_set1_groups_list`", which also seems to be the outcome of the [discussion here](https://github.com/openssl/openssl/issues/11548).

So, unless there is dissent I think the best action would be to withdraw/close this proposal, thanking you all for your feedback/input!

**Comment by t8m** [comment]
> This would (still) necessitate adding the `default_group` bit to `TLS_GROUP_INFO` also in the internal providers. Do I understand you right? Or do you suggest inserting any group on the "preseed list" regardless of whether or not it has a `default_group` bit set (as it is a "privileged prioritized group" to begin with)?

Yes, the built-in providers would have to add the default_group bit as well.

I disagree it would be too complicated but anyway staying with "if the currently built-in choice and priority of default groups doesn't work for you, use SSL_set1_groups_list" is also OK.

**Comment by mattcaswell** [comment]
What should happen in the event that, an explicit group list has not been set, there are no groups from the built-in choice available, but we do have other groups available? I guess the handshake fails?

**Comment by baentsch** [comment]
I'm tempted to simply answer "Yes", but I'm not sure I fully understand the scenario you describe, so let me expand: 

What do you mean by "we have other groups available"? Available via a provider but not registered in the `supported_groups` structure (either added "manually" via `SSL_set1_groups_list` or automatically via the `supported_groups_default` struct discussed here)? In that case, I'd think [RFC 8446 section 4.1.1](https://tools.ietf.org/html/rfc8446#section-4.1.1) rules (and assuming we're not talking about PSK as an alternative): 
> If there is no overlap between the received "supported_groups" and the groups supported by the server, then the server MUST abort the handshake with a "handshake_failure" or an "insufficient_security" alert

So indeed, it could now be possible to configure OpenSSL such that it would have provider(s) that have the crypto to establish a connection but connection establishment will still fail because they client and server will never know about this fact (again, if `SSL_set1_groups_list` has not been explicitly called specifying now available groups). Not even a [Hello Retry Request](https://tools.ietf.org/html/rfc8446#section-4.1.4) would be permissible as the spec states
> a HelloRetryRequest MUST NOT contain any extensions that were not first offered by the client in its ClientHello

But then again, the only way out of this would be to add _all_ groups that _any_ provider offers automatically to the `supported_groups` structure (possibly subject to some preference ordering). Certainly the client would _have_ to send all of them to ensure the server has a chance to react. This would go beyond the proposal of this issue (but clearly is even more simple to implement :-).

**Comment by mattcaswell** [comment]
> What do you mean by "we have other groups available"? Available via a provider but not registered in the supported_groups structure (either added "manually" via SSL_set1_groups_list or automatically via the supported_groups_default struct discussed here)? In that case, I'd think RFC 8446 section 4.1.1 rules (and assuming we're not talking about PSK as an alternative):

Let me give an example. Lets assume an application acting as a client *only* loads the oqs provider (not default, or fips etc), but has not explicitly called `SSL_set1_groups_list` or similar. In that case there are no groups in `supported_groups_default` that are actually available, so when the application comes to construct the ClientHello the supported_groups extension will be empty. This is actually invalid.

We could either:
1) Fail the handshake immediately (before it has even started)
or
2) Continue, but not send the extension

In the latter case, if no PSK is being sent, then it will not be possible to negotiate TLSv1.3 at all. Therefore, if we were going to go down that route, then we should probably detect this and automatically disable TLSv1.3 (i.e. so that supported_versions etc is not sent). This automatic detection of whether or not we should have TLSv1.3 enabled seems quite complicated to me, and therefore I prefer option (1). This has some precedent: if you don't configure any TLSv1.3 ciphersuites and attempt to start a TLSv1.3 handshake it will fail immediately.

One slight issue with this approach is with the "no-ec" and "no-dh" case. In 1.1.1 if you specify "no-ec" and "no-dh" then TLSv1.3 is automatically disabled via Configure. Due to groups now being pluggable we've already said that this means we should stop doing that (I have a PR in the works). But, taking that thought further that means if you do configure with "no-ec" and "no-dh" then the default provider will not have any support for TLS groups at all. Therefore, unless you load some other provider, you will again hit the situation where no groups in "supported_groups_default" are actually available. If we go with option (1) then this means the handshake fails. So configuring with "no-ec" and "no-dh" means that by default all TLSv1.3 connections will immediately fail without further config. This is different to the behaviour in 1.1.1 where configuring "no-ec" and "no-dh" just meant that TLSv1.3 was never attempted. This is perhaps ok (who wants to configure with "no-ec" and "no-dh" anyway - and if you do you must live with the consequences)

**Comment by baentsch** [comment]
> I have a PR in the works

Thanks for letting me know you're working on this: The "no-ec, no-dh" case was precisely what made me raise this issue: When trying a PR for this myself I ran into this very problem (of an empty default list) in testing:

```
        # INFO:  @ test/helpers/ssltestlib.c:960
        # SSL_accept() failed -1, 1
        # C0A1209AED7F0000:error:0A0000D1:SSL routines:tls_parse_ctos_key_share:missing supported groups extension:ssl/statem/extensions_srvr.c:615:
        # INFO:  @ test/helpers/ssltestlib.c:942
        # SSL_connect() failed -1, 1
        # C0A1209AED7F0000:error:0A000455:SSL routines:ssl3_read_bytes:tlsv13 alert missing extension:ssl/record/rec_layer_s3.c:1584:SSL alert number 109
```

But to answer the question: If we don't adapt the default list _somehow_ "immediate failure" is what already happens (in an oqsprovider, no-ec, no-dh, no-default provider config).

**Comment by t8m** [comment]
But this is incorrect failure :grin: as the failure should happen already on the client side - the current code will produce TLS-1.3 non-conforming client hello which is then correctly rejected server side.

**Comment by t8m** [comment]
I still do not think my proposal would be that bad to implement, I might look at it later.

**Comment by mattcaswell** [comment]
> I still do not think my proposal would be that bad to implement, I might look at it later.

I'd suggest waiting until at least my PR for supporting "no-ec" and "no-dh" is done. Its touching a lot of the same areas.

**Comment by baentsch** [comment]
@romen : You stated

> edit the global OPENSSLDIR/openssl.cnf (openssl version -d) or the per-application configuration file set 
> using the OPENSSL_CONF env var to:
>     - load the OQS provider
>     - add whatever KEM the user wants to support by default to the default ssl configuration for server and/or client

It's clear to me how to add (and activate) oqsprovider in the openssl.cnf file (item 1 above) but I don't know how to set the default groups list in that file (item 2 above): Can you please point to documentation (or a sample config) so I can try this?

@mattcaswell : You stated

> I'd suggest waiting until at least my PR for supporting "no-ec" and "no-dh" is done. Its touching a lot of the same areas.

--> Would it be OK to carry this further now? Most of the "integrating quantum-safe KEM" into openssl-using applications is in enabling/correcting calls to `SSL_set1_groups_list` (the latest in [openvpn](https://github.com/OpenVPN/openvpn/pull/170)): It would be nice to have a more general mechanism driven by OpenSSL itself instead.

**Comment by romen** [comment]
> @romen : You stated
> 
> > edit the global OPENSSLDIR/openssl.cnf (openssl version -d) or the per-application configuration file set
> > using the OPENSSL_CONF env var to:
> > - load the OQS provider
> > - add whatever KEM the user wants to support by default to the default ssl configuration for server and/or client
> 
> It's clear to me how to add (and activate) oqsprovider in the openssl.cnf file (item 1 above) but I don't know how to set the default groups list in that file (item 2 above): Can you please point to documentation (or a sample config) so I can try this?

It would look similar to this:

~~~
openssl_conf = openssl_init

[openssl_init]
ssl_conf = ssl_sect

[ssl_sect]
system_default = system_default_sect

[system_default_sect]
Groups = SNTRUP761:SNTRUP857:X25519:P-256
~~~

**Comment by mattcaswell** [comment]
> Would it be OK to carry this further now? 

Sure. I think the PR I was referring to was #13916 which was merged long ago.

**Comment by baentsch** [comment]
> It would look similar to this:

@romen : Works like a charm: Thanks! FYI, this is now part of an [OpenVPN-OpenSSL3-OQS integration via docker](https://github.com/open-quantum-safe/oqs-demos/pull/135).

> Sure. I think the PR I was referring to was https://github.com/openssl/openssl/pull/13916 which was merged long ago.

@mattcaswell : Thanks for that reconfirmation. Given how many different quantum-safe algorithms (and hybrid combinations) are still in "the running" of the [NIST PQC competition](https://csrc.nist.gov/projects/post-quantum-cryptography) it's probably better to wait a bit more on this, though: Adding all to the KEM groups list via the provider is just unwieldy/inefficient -- and the `openssl.cnf` approach above resolves the issue very well for now. NIST should whittle down the list soon, so we could revisit this then.
